### PR TITLE
New command 'list_clients'

### PR DESCRIPTION
**Pull request by t-wissmann**
For the new command 'list_clients' the main challenge was to extend
ArgParse by command line flags that take arguments. To make the
completion work, I also had to extend the completion objects by a
'prepend' string that is prepended to every completion result.

This implements the main feature request of #831 and also includes the
functionality of the rejected PR #1249.

### CODE REVIEW

**Comment by t-wissmann** [comment]
I'm not sure whether `--in-tag=…` or `--on-tag=…` is better (and same for `--on-monitor=…`); "in" sounds a bit odd, but would be more consistent with `--in-frame`

**Comment by tiosgz** [comment]
> I'm not sure whether `--in-tag=…` or `--on-tag=…` is better (and same for `--on-monitor=…`); "in" sounds a bit odd, but would be more consistent with `--in-frame`

(or remove the prepositions… though @etosan said it sounds more natural to him with them)

File: tests/test_global_commands.py [hunk 1]
```diff
@@ -264,3 +264,115 @@ def test_focus_nth_completion(hlwm):
...
+    def list_clients(flags):
+        return sorted(hlwm.call(['list_clients'] + flags).stdout.splitlines())
+
+    assert list_clients(['--in-frame=']) == sorted(clients)
+    assert list_clients(['--in-frame=@']) == sorted(clients[0:2])
+    assert list_clients(['--in-frame=0']) == sorted(clients[0:2])
+    assert list_clients(['--in-frame=1']) == sorted(clients[2:4])
+
+
+def test_list_clients_invalid_arg(hlwm):
```
Comment (blarz) [inline tests/test_global_commands.py:331]: I tested ```list_clients --on-monitor=``` and it works, is this intended?

File: src/argparse.cpp [hunk 2]
```diff
@@ -213,17 +234,55 @@ void ArgParse::completion(Completion& complete)
 }
 
 /**
- * @brief try parse a flag
+ * @brief try parse a flag, possibly throwing an exception
```
Comment (blarz) [inline src/argparse.cpp:237]: try *to* parse a flag?

**Comment by blarz** [comment]
> > I'm not sure whether `--in-tag=…` or `--on-tag=…` is better (and same for `--on-monitor=…`); "in" sounds a bit odd, but would be more consistent with `--in-frame`
> 
> (or remove the prepositions… though @etosan said it sounds more natural to him with them)

I wouldn't oppose `--tag` or `--monitor` either to be honest.

File: tests/test_global_commands.py (hunk 1 above)
Comment (t-wissmann) [reply tests/test_global_commands.py:331]: Yes, this is a corner case because the empty string refers to the focused monitor.

File: src/argparse.cpp (hunk 2 above)
Comment (t-wissmann) [reply src/argparse.cpp]: [x] fixed, thanks!

**Comment by t-wissmann** [comment]
Yes, simply `--tag=` and `--monitor=` make sense. I've renamed them. So far, no hlwm commands follow the `--in-…`/`--or-…` pattern (the only command line flags with parameters so far were the ones for the `rule` command)

File: src/argparse.cpp [hunk 3]
```diff
@@ -54,7 +54,21 @@ bool ArgParse::parsingFails(Input& input, Output& output)
...
         // try to parse this token as a flag
-        while (optionalArgumentsRemaining && tryParseFlag(valueString)) {
+        while (optionalArgumentsRemaining) {
+            try {
+                if (!tryParseFlag(valueString)) {
+                    // if it is not a flag, then break, i.e
+                    // continue with positional args
+                    break;
+                }
+            }  catch (std::exception& e) {
```
Comment (blarz) [inline src/argparse.cpp:64]: This branch is not covered according to the coverage report. Does it make sense to add a test for this?

File: src/argparse.cpp (hunk 2 above)
Comment (blarz) [reply src/argparse.cpp]: Thanks!

File: tests/test_global_commands.py (hunk 1 above)
Comment (blarz) [reply tests/test_global_commands.py:331]: I see!

File: src/argparse.cpp (hunk 3 above)
Comment (t-wissmann) [reply src/argparse.cpp:64]: This code is not reachable at the moment because it is only executed if there is a command line flag with parameters in a command with position arguments. `list_clients` however is the onle command with parametrized flags, but it does not have positional args. Once this is merged, I plan to extend `foreach` by a `--name-match=…` flag, and then we can test this branch.

Comment (blarz) [reply src/argparse.cpp:64]: Ok thanks! Just wanted to check :)

**Review by blarz — APPROVED** [review]
Very nice, thanks!
